contracts/contracts/gatekeepers/EASGatekeeper.sol

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { SignUpGatekeeper } from "./SignUpGatekeeper.sol";
+
+/// @notice A struct representing a single attestation.
+struct Attestation {
+  bytes32 uid; // A unique identifier of the attestation.
+  bytes32 schema; // The unique identifier of the schema.
+  uint64 time; // The time when the attestation was created (Unix timestamp).
+  uint64 expirationTime; // The time when the attestation expires (Unix timestamp).
+  uint64 revocationTime; // The time when the attestation was revoked (Unix timestamp).
+  bytes32 refUID; // The UID of the related attestation.
+  address recipient; // The recipient of the attestation.
+  address attester; // The attester/sender of the attestation.
+  bool revocable; // Whether the attestation is revocable.
+  bytes data; // Custom attestation data.
+}
+
+interface IEAS {
+  function getAttestation(bytes32 uid) external view returns (Attestation memory);
+}
+
+contract EASGatekeeper is SignUpGatekeeper, Ownable {
+  // the reference to the EAS contract
+  IEAS private immutable eas;
+
+  // the schema to check against
+  bytes32 private immutable schema;
+
+  // a mapping of attestations that have already registered
+  mapping(bytes32 => bool) public registeredAttestations;
+
+  /// @notice the reference to the MACI contract
+  address public maci;
+
+  /// @notice custom errors
+  error AttestationRevoked();
+  error AlreadyRegistered();
+  error NotYourAttestation();
+  error InvalidAttestation();
+  error OnlyMACI();
+
+  /**
+   * Deploy an instance of EASGatekeeper
+   * @param _eas The EAS contract
+   * @param _schema The schema UID
+   */
+  constructor(address _eas, bytes32 _schema) payable Ownable() {
+    eas = IEAS(_eas);
+    schema = _schema;
+  }
+
+  /// @notice Adds an uninitialised MACI instance to allow for token signups
+  /// @param _maci The MACI contract interface to be stored
+  function setMaciInstance(address _maci) public override onlyOwner {
+    maci = _maci;
+  }
+
+  /**
+   * Register an user based on their attestation
+   * @dev Throw if the attestation is not valid or just complete silently
+   * @param _user The user's Ethereum address.
+   * @param _data The ABI-encoded schemaId as a uint256.
+   */
+  function register(address _user, bytes memory _data) public override {
+    bytes32 attestationId = abi.decode(_data, (bytes32));
+
+    if (maci != msg.sender) revert OnlyMACI();
+
+    if (registeredAttestations[attestationId]) revert AlreadyRegistered();
+
+    registeredAttestations[attestationId] = true;
+    Attestation memory attestation = eas.getAttestation(attestationId);
```

Comment:
Should we also add a check that the `attester` of the `attestation` is a valid authority?

e.g. in the case of OP RPGF, we only care about attestations from the OP foundation, right?

Not sure if that's already accounted for, but maybe we need to add e.g. an authority to the constructor where we only register if the attestation came from the authority.

Reply:
In the constructor we pass the schema, that is the schema created by OP foundation, so we check that the user attestation is part of this schema (in the test cases is the badgeholder schema).

Can someone create an attestation from someone else schema? like you come and give an attestation of OP foundation schema to someone? If so, then yes we should add that, but if you can't (which it just makes sense that you can't as that sounds dodgy), then what we do here I think is enough.

Reply:
oh I'll take that back, it looks like someone can create an attestation from someone else schema.. still sounds weird in my mind, probably didn't really understand how EAS work 😆

Reply:
@samajammin check added